File: backend/src/main/java/com/roommate/hub/entity/RoomPost.java
package com.roommate.hub.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;

@Entity
@Table(name = "room_posts")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class RoomPost {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String description;

    @Column(nullable = false)
    private Double price;

    @Column(nullable = false, length = 255)
    private String address;

    @Column(length = 100)
    private String district;

    private Double deposit;

    @Column(name = "electricity_water_cost")
    private Double electricityWaterCost;

    private Double area;

    @Column(nullable = false)
    private Integer maxOccupants;

    @Column(nullable = false)
    @Builder.Default
    private Integer currentOccupants = 0;

    @Column(length = 500)
    private String amenities;

    @Column(length = 255)
    private String imageUrl;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private PostStatus status; // AVAILABLE, FILLED

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    public enum PostStatus {
        PENDING,    // Chờ Admin duyệt
        APPROVED,   // Đã duyệt (hiển thị công khai)
        REJECTED,   // Bị từ chối
        AVAILABLE   // Tương đương APPROVED
    }
}

File: frontend/lib/models/room_post.dart
class RoomPost {
  final int id;
  final String title;
  final String description;
  final double price;
  final String address;
  final int maxOccupants;
  final String authorName;
  final int authorId;
  final String? imageUrl;
  final double? areaM2;
  final List<String> amenities;

  RoomPost({
    required this.id,
    required this.title,
    required this.description,
    required this.price,
    required this.address,
    required this.maxOccupants,
    required this.authorName,
    required this.authorId,
    this.imageUrl,
    this.areaM2,
    this.amenities = const <String>[],
  });

  factory RoomPost.fromJson(Map<String, dynamic> json) {
    return RoomPost(
      id: json['id'] as int? ?? 0,
      title: json['title'] as String? ?? '',
      description: json['description'] as String? ?? '',
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      address: json['address'] as String? ?? '',
      maxOccupants: json['maxOccupants'] as int? ?? 1,
      authorName: json['authorName'] as String? ?? 'Ẩn danh',
      authorId: json['authorId'] as int? ?? 0,
      imageUrl: (json['imageUrl'] ?? json['image_url']) as String?,
      areaM2: (json['areaM2'] ?? json['area_m2'] ?? json['area']) is num
          ? ((json['areaM2'] ?? json['area_m2'] ?? json['area']) as num)
                .toDouble()
          : null,
      amenities:
          (json['amenities'] as List<dynamic>?)?.whereType<String>().toList(
            growable: false,
          ) ??
          const <String>[],
    );
  }
}
